### PR TITLE
Nk/interlock tele

### PR DESCRIPTION
**Pull request by nkrackow**
This PR
* adds the laser interlock state to the telemetry. 
* refactors the laser interlock to store information about the tripping reason.
* refactors the logic of the thermostat interlock such that it can be reset by resetting the laser interlock.

### CODE REVIEW

**Comment by nkrackow** [comment]
The clippy lints at the end are due to commented out parts that don't work without Driver HW right now.

File: src/hardware/driver/mod.rs [hunk 1]
```diff
@@ -57,17 +58,37 @@ impl ChannelVariant {
...
+
+    pub fn set(&mut self, state: Option<LaserInterlockTripped>) {
         match state {
-            PinState::High => Self::NotTripped,
-            PinState::Low => Self::Tripped,
+            Some(LaserInterlockTripped::Reset) => self.pin.set_low(),
+            Some(LaserInterlockTripped::Thermostat) => self.pin.set_low(),
+            Some(LaserInterlockTripped::Overcurrent(_)) => self.pin.set_low(),
+            Some(LaserInterlockTripped::Overvoltage(_)) => self.pin.set_low(),
+            None => self.pin.set_high(),
```
Comment (jordens) [inline src/hardware/driver/mod.rs:90]:
```suggestion
            self.pin.set_state(state.is_none().into());
```

File: src/bin/driver.rs [hunk 2]
```diff
@@ -417,6 +415,34 @@ mod app {
...
         });
 
+        // Only set interlock pin high if all channels disabled and false -> true transition.
+        if !old_settings.reset_laser_interlock
+            && new_settings.reset_laser_interlock
+        {
+            if c.shared.output_state.lock(|output| {
+                output[driver::Channel::HighPower as usize].is_enabled()
+            }) {
+                log::error!("Cannot reset laser interlock while HighPower channel is enabled.")
```
Comment (jordens) [inline src/bin/driver.rs:425]: Always good idea to tell the user what to do here.

File: src/hardware/driver/mod.rs [hunk 3]
```diff
@@ -57,17 +58,37 @@ impl ChannelVariant {
...
+pub enum LaserInterlockTripped {
     #[default]
-    Tripped,
+    Reset, // Tripped after device reset
+    Thermostat,
+    Overcurrent(Channel),
+    Overvoltage(Channel),
+}
+pub struct LaserInterlock {
+    pub state: Option<LaserInterlockTripped>,
```
Comment (jordens) [inline src/hardware/driver/mod.rs:69]: Don't make these `pub`!
Better do `pub fn state(&self) -> ...`.

Comment (jordens) [inline src/hardware/driver/mod.rs:69]: And better call this `reason`.
Call the `enum` `Reason`.

File: src/bin/driver.rs [hunk 4]
```diff
@@ -600,20 +616,21 @@ mod app {
...
         for (i, (interlock_current, read)) in interlock_current
             .iter()
             .zip((current_reads).iter())
             .enumerate()
         {
             if (read > interlock_current) & interlock_asserted {
-                c.shared.laser_interlock_pin.lock(|pin| pin.set_low());
+                let ch = Channel::try_from(i).unwrap();
+                c.shared.laser_interlock.lock(|ilock| {
+                    ilock.set(Some(LaserInterlockTripped::Overcurrent(ch)))
```
Comment (jordens) [inline src/bin/driver.rs:629]: `interlock_asserted`  becomes inconsistent with the pin and and `ilock.state` becomes inconsistent with the state that caused the interlock to fire!

**Review by jordens — CHANGES_REQUESTED** [review]

File: src/hardware/driver/mod.rs (hunk 1 above)
Comment (nkrackow) [reply src/hardware/driver/mod.rs:90]: nice..

File: src/bin/driver.rs (hunk 4 above)
Comment (nkrackow) [reply src/bin/driver.rs:629]: Ok, but does that matter if `interlock_asserted` is dropped anyways at the end of the task? If we have an overvoltage condition as well that would trp the interlock a second time, but that isn't a problem either right?

**Pull request merged**